qiskit-superstaq/qiskit_superstaq/superstaq_job.py

```diff
@@ -74,13 +112,19 @@ def result(self, timeout: Optional[float] = None, wait: float = 5) -> qiskit.res
         """
         timeout = timeout or self._backend._provider._client.max_retry_seconds
         results = self._wait_for_results(timeout, wait)
+        circ_meas_list: List[List[int]] = []
 
         # create list of result dictionaries
         results_list = []
-        for result in results:
+        for index, result in enumerate(results):
             counts = result["samples"]
             if counts:  # change endianess to match Qiskit
+                num_clbits = self._backend._num_clbits_in_circ[index]
```

Comment:
one problem i just realized: it's not necessarily the case that `self._backend` is the same backend instance that created the job; for example, as long as i save my job_id, the following should work even is i restart my python shell after submitting the job:
```python
provider = qss.SuperstaqProvider()
SuperstaqJob(provider.get_backend(target), job_id)`
```
(relatedly, does anything need to change if i create multiple jobs using the same backend before retrieving any of their results?)


given this, afaict it seems like we need to make one of two server-side changes to make this work (unless you see another workaround):
1. we could have the server return the input circuits alongside the compiled ones as part of the job info, so they're always available regardless of how the job was constructed)
2. we could handle more of this server-side by making sure the compiled circuit's clbits are always the same as the input circuits'

imo we should eventually aim to do both of these, though (2) might be a bigger change so it seems reasonable to start with just (1) if that makes sense to you

Reply:
So yes that's a good catch, this implementation would fail if you were to instantiate `provider.get_backend()` at a later point. And to your other question, I think I overlooked the scenario where multiple jobs are created but the results are retrieved at a later time so that is also something to fix. 

As a client-side fix, we could store `num_clbits` within the circuit metadata itself during the serialization process (which is unaffected during server-side compilation) and then can always be retrieved within `SuperstaqJob` with the `compiled_circuits()` method. And then, after deserialization, this information could be stripped from the circuit metadata into a job attribute so that the altered circuit metadata does not bewilder the user. Other than that workaround, nothing strikes me at the moment.

Though, in the beginning, I did start out wanting to do the (1) suggestion, but stopped after realizing the further server-side changes (and thinking client-side might prove sufficient). However, if it's something that would be implemented eventually anyways, I could definitely start with (1).

Reply:
> As a client-side fix, we could store `num_clbits` within the circuit metadata...

oh cool, as a temporary fix that's pretty nice

> ...if it's something that would be implemented eventually anyways, I could definitely start with (1).

i definitely think it's something we want, but depending on your bandwidth/how much you want to prioritize this imo the fix you outlined above is pretty sufficient for the short term

Reply:
With the inclusion of the `_get_circuits("input_circuit")` method, `num_clbits_in_circ` is no longer associated with a `SuperstaqBackend` and can be retrieved within a `SuperstaqJob` itself.